### ISSUE TITLE
Allow refinement of wildcard bounds

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#260]
```
What steps will reproduce the problem?
The following code fails to typecheck:

import checkers.nullness.quals.*;
import java.util.List;

class TestGenerics {
    void test(@NonNull List<? extends @Nullable String> list1,  
            @NonNull List<? extends @NonNull String> list2) {  

        list1 = list2;
       @NonNull String s = list1.get(0);
    } 
}


What is the expected output? What do you see instead?
No errors are expected, this error is printed:

TestGenerics.java:17: error: incompatible types in assignment.
        @NonNull String s = list1.get(0);
                                     ^
  found   : ? extends @Initialized @Nullable String
  required: @UnknownInitialization @NonNull String
1 error

What version of the product are you using? On what operating system?
javac 1.8.0-jsr308-1.6.7
Checker-framework: 7029:6bedff1c764c

Please provide any additional information below.

We refine main qualifiers on assignment, but it should be possible to refine 
the qualifiers of a wildcard bound given the wildcard subtyping rules here: 
http://docs.oracle.com/javase/tutorial/java/generics/subtyping.html
```

Original issue reported on code.google.com by `mcart...@cs.washington.edu` on 4 Oct 2013 at 1:18

**Comment by smillst** [comment]
Duplicate of #470.  Closed in favor of #470.
